File: Ruby/quick_sort.rb
def quick_sort(list)
  raise 'list must be array or something collection' unless list.kind_of?(Array)
  case list.length
    when 0 then
      []
    else
      head, *tail = list
      smaller = quick_sort(tail.select { |x| x <= head })
      larger = quick_sort(tail.select { |x| x >= head })
      (smaller.concat([head])).concat(larger)
  end
end


puts(quick_sort([5,4,2,1,77,8,555,44444]))

File: TypeScript/quicksort.ts
var quickSort = (array : number[]):number[] => {
  switch(array.length){
    case 0:
      return [];
    default:
      var [head, ...tail] = array;
      var smaller = quickSort(tail.filter(x => x <= head));
      var greater = quickSort(tail.filter(x => x >= head));
      return smaller.concat([head]).concat(greater);
  }
};

console.log(quickSort([1,5,77,2,3,11,44]))
